### PR TITLE
Wrap text pull request

### PR DESCRIPTION
**Pull request by ktei**
@maxkfranz @kurniaikhwan 
This is the new text wrapping feature pull request, which should have fixed the glitches in previous pull request.
Note that for a better looking graph, we do not wrap text for edge label text.
capture

### CODE REVIEW

**Comment by kurniaikhwan** [comment]
@ktei , how to implement your code to my cytoscape's network view?

i really hard to understand :D

**Comment by ktei** [comment]
@kurniaikhwan , the easiest way for you to do this (for now) is:
0. fork this cytoscape project
1. git co unstable (switch to unstable branch)
2. check this pull request. there should be 3 files changed and you could just manually copy paste those lines into your cytoscape. Then you need to run gulp build to build your own cytoscape

**Comment by kurniaikhwan** [comment]
@maxkfranz 
please put this feature in to the cytoscape desktop as soon as possible. :D

**Comment by ktei** [comment]
@kurniaikhwan I don't think the project owner is also maintaining the desktop version. And for this pull request, it is actually about text wrapping, not label rotating.

**Comment by maxkfranz** [comment]
Merged; will make a couple changes regarding performance shortly...

**Comment by maxkfranz** [comment]
This change greatly improves the wrapped labels implementation.  It has a lot of caching now, and it should be less buggy now.

There were problems with label dimensions for fitting etc. because `cxt.measureText()` was used.  There was also the problem that top aligned labels wouldn't align properly anymore.  These and other issues have been fixed by the commit.

**Comment by maxkfranz** [comment]
Example of labels above and below: screen shot 2015-03-10 at 16 51 55

**Comment by maxkfranz** [comment]
@ktei  Is there a reason to continue to disable this for edges?  We probably should just allow it unless it causes some problem.

**Comment by ktei** [comment]
@maxkfranz I merged your changes but found a bug:
See the following screenshot:
capture
Note that all the labels show 'cytoscape rocks! cytoscape' but in fact the label content is 'cytoscape rocks! cytoscape rocks!' so all the labels miss the last word.

This is my html:
```html
<!DOCTYPE html>
<html>
<head>
    <style>
        body {
            font: 14px helvetica neue, helvetica, arial, sans-serif;
        }

        #cy {
            height: 100%;
            width: 100%;
            position: absolute;
            left: 0;
            top: 0;
        }

        #eat {
            position: absolute;
            left: 1em;
            top: 1em;
            font-size: 1em;
            z-index: -1;
            color: #c88;
        }
    </style>
    <meta charset=utf-8 />
    <title>Cytoscape.js initialisation</title>
    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1/jquery.min.js"></script>
    <script src="~/Vendor/cytoscape.js/build/cytoscape.js"></script>
</head>
<body>
    <div id="cy"></div>
    <script>
        $(function () { // on dom ready

            // photos from flickr with creative commons license

            var cy = cytoscape({
                container: document.getElementById('cy'),

                style: cytoscape.stylesheet()
                  .selector('node')
                    .css({
                        'height': 80,
                        'width': 80,
                        'background-fit': 'cover',
                        'border-color': '#000',
                        'border-width': 3,
                        'border-opacity': 0.5,
                        'text-valign': 'bottom',
                        'text-wrap': 'wrap',
                        'content': 'cytoscape rocks! cytoscape rocks!',
                    })
                  .selector('.eating')
                    .css({
                        'border-color': 'red'
                    })
                  .selector('.eater')
                    .css({
                        'border-width': 9
                    })
                  .selector('edge')
                    .css({
                        'color': '#333',
                        'width': 6,
                        'content': 'animal',
                        'target-arrow-shape': 'triangle',
                        'line-color': '#ffaaaa',
                        'target-arrow-color': '#ffaaaa',
                        'edge-text-rotation': 'autorotate',
                        'text-background-color': '#ffaaaa',
                        'text-border-color': 'red'
                    })
                  .selector('edge:selected')
                    .css({
                        'text-background-color': 'red'
                    })
                  .selector('#bird')
                    .css({
                        'background-image': 'https://farm8.staticflickr.com/7272/7633179468_3e19e45a0c_b.jpg'
                    })
                  .selector('#cat')
                    .css({
                        'background-image': 'https://farm2.staticflickr.com/1261/1413379559_412a540d29_b.jpg'
                    })
                  .selector('#ladybug')
                    .css({
                        'background-image': 'https://farm4.staticflickr.com/3063/2751740612_af11fb090b_b.jpg'
                    })
                .selector('#aphid')
                    .css({
                        'background-image': 'https://farm9.staticflickr.com/8316/8003798443_32d01257c8_b.jpg'
                    })
                .selector('#rose')
                    .css({
                        'background-image': 'https://farm6.staticflickr.com/5109/5817854163_eaccd688f5_b.jpg'
                    })
                .selector('#grasshopper')
                    .css({
                        'background-image': 'https://farm7.staticflickr.com/6098/6224655456_f4c3c98589_b.jpg'
                    })
                .selector('#plant')
                    .css({
                        'background-image': 'https://farm1.staticflickr.com/231/524893064_f49a4d1d10_z.jpg'
                    })
                .selector('#wheat')
                    .css({
                        'background-image': 'https://farm3.staticflickr.com/2660/3715569167_7e978e8319_b.jpg'
                    }),

                elements: {
                    nodes: [
                      { data: { id: 'cat' } },
                      { data: { id: 'bird' } },
                      { data: { id: 'ladybug' } },
                      { data: { id: 'aphid' } },
                      { data: { id: 'rose' } },
                      { data: { id: 'grasshopper' } },
                      { data: { id: 'plant' } },
                      { data: { id: 'wheat' } }
                    ],
                    edges: [
                      { data: { source: 'cat', target: 'bird' } },
                      { data: { source: 'bird', target: 'ladybug' } },
                      { data: { source: 'bird', target: 'grasshopper' } },
                      { data: { source: 'grasshopper', target: 'plant' } },
                      { data: { source: 'grasshopper', target: 'wheat' } },
                      { data: { source: 'ladybug', target: 'aphid' } },
                      { data: { source: 'aphid', target: 'rose' } }
                    ]
                },

                layout: {
                    name: 'cose',
                    directed: true,
                    padding: 10
                }
            }); // cy init
        });
    </script>
</body>
</html>
```

I think the problem is in `renderer.canvas.coord-ele-math.js`:
```javascript
            if( testW <= maxW ){ // word fits on current line
              subline += word + ' ';
            } else { // word starts new line
              wrappedLines.push( subline );
              subline = word + ' ';
            }
```
The code above has a bit issue. One of the broken cases is that:
Suppose the label content is `cytoscape rocks! cytoscape rocks!`.
What if you are detecting the last word, which is `rocks!` however because testW <= maxW already, it just jumps to `wrappedLines.push(subline)`, which pushes `cytoscape rocks! cytoscape`, and that's it, now we have already missed the last `rocks!`.

I tried to fix that but haven't really found a proper way to do that. Could you check this and try to fix it? Thanks.

**Comment by ktei** [comment]
@maxkfranz 
I double checked my original change and I think this issue can be fixed like this:
```javascript
          var words = line.split(/\s+/); // NB: assume collapsed whitespace into single space
          var subline = '';
          for( var w = 0; w < words.length; w++ ){
            var word = words[w];
            var testLine = subline + word + ' ';
            var testDims = this.calculateLabelDimensions( ele, testLine, 'testLine=' + testLine );
            var testW = testDims.width;

            if (testW > maxW && w > 0) {
              wrappedLines.push(subline);
              subline = words[w] + ' ';
            } else {
              subline = testLine;
            }

            // if( testW <= maxW ){ // word fits on current line
            //   subline += word + ' ';
            // } else { // word starts new line
            //   wrappedLines.push( subline );
            //   subline = word + ' ';
            // }
          }
          wrappedLines.push(subline);
```

**Comment by maxkfranz** [comment]
I think this may be a bit safer.

**Comment by ktei** [comment]
thanks @maxkfranz 
I think this is working fine now.

**Comment by ktei** [comment]
As for your last question: 'Is there a reason to continue to disable this for edges?'
I have no problem with that but my only concern is that wrapped text might make the edge text rectangle bigger, so mostly I'd avoid wrapping that. But theoretically, it's nothing wrong to wrap edge label text. Cheers

**Comment by weirdalsuperfan** [comment]
So is this implemented in the normal cytoscape.js, or is this still a separate branch? Under http://js.cytoscape.org/#style/labels there's a section about text wrapping, which is why I ask.
